kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -521,45 +497,45 @@ func (c *loadBalancerController) syncService(svcObj serviceObject) error {
 }
 
 // needsIPsAssigned determines if service IPFamilyPolicy is requirement is fulfilled by number of assigned IPs in IPAM storage
-func (c *loadBalancerController) needsIPsAssigned(svc *v1.Service, svcObj serviceObject) bool {
+func (c *loadBalancerController) needsIPsAssigned(svc *v1.Service, svcKey serviceKey) bool {
 	switch *svc.Spec.IPFamilyPolicy {
 	case v1.IPFamilyPolicySingleStack:
-		return len(c.allocationTracker.ipsByService[svcObj]) == 1
+		return len(c.allocationTracker.ipsByService[svcKey]) != 1
 	case v1.IPFamilyPolicyRequireDualStack, v1.IPFamilyPolicyPreferDualStack:
-		return len(c.allocationTracker.ipsByService[svcObj]) == 2
+		return len(c.allocationTracker.ipsByService[svcKey]) != 2
```

Comment:
This isn't a perfect check - ideally we'd also check that for dual stack services, the IPs in our IPAM DB are actually dual-stack.

Otherwise, we might return "true" here when there are two IPv4 IPs in our database (which technically necessitates a new IPv6 address).

this isn't super urgent though - I'd expect that our code would spot the extra IPv4, release it, and then the next sync we'd fail this check.